### PR TITLE
Implements 3d text feature for engine

### PR DESCRIPTION
**Pull request by hobogenized**
- Contains stubs for 2d text for when it will be added
- Allows for text to be added to scenes

### CODE REVIEW

**Comment by acgaudette** [comment]
Reason for new dependencies: `rand`, `image`?

**Comment by acgaudette** [comment]
Getting a `DEBUG_REPORT: Object: VK_NULL_HANDLE (Type = 0) | Type mismatch on descriptor slot 0.1 (used as type 'ptr to const uniform sampler+image(dim=1, sampled=1)') but descriptor of type VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC`

**Comment by acgaudette** [comment]
Re: component names. How about `label` for text2d and `text` for text3d? I'm not sold, but might be better than what we have. jw

File: src/font.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    pub uv_height: f32,
+    pub char_map: fnv::FnvHashMap<i32, Bmchar>,
+}
+
+fn get_value_from_pair<'a> (pair: &'a str) -> &'a str {
+    let val:Vec<&str> = pair.split('=').collect();
+    val[1]
+}
+
+fn parse_bmchar<'a>(full_str: &'a str) -> Bmchar {
```
Comment (acgaudette) [inline src/font.rs]: This is a good place for proper error handling + maybe a variable macro.

File: src/font.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        let mut uv_height = 0f32;
+        let mut base_width = 0f32;
+        let mut line_height = 0f32;
+
+        for line in file.lines() {
+            let line_string = line.unwrap();
+            let clone = line_string.clone();
+            let mut iter = clone.split_whitespace();
+            match iter.next() {
+                Some("info") => {
```
Comment (acgaudette) [inline src/font.rs]: Can get rid of the long comment. Maybe could use the info `face` at least to log the font we're loading.

File: src/components/mod.rs [hunk 3]
```diff
@@ -40,4 +42,5 @@ pub struct Container {
     pub lights:     light::Manager,
     pub draws:      draw::Manager,
     pub softbodies: softbody::Manager,
+    pub text3ds:    text3d::Manager,
```
Comment (acgaudette) [inline src/components/mod.rs]: Re: component names. How about label for text2d and text for text3d? I'm not sold, but might be better than what we have. jw

File: Cargo.toml [hunk 4]
```diff
@@ -10,3 +10,5 @@ fnv = "1.0.3"
 rust-ini = "0.10.0"
 lazy_static = "1.0.0"
 tobj = "0.1.6"
+rand = "0.5.0"
+image = "0.19.0"
```
Comment (acgaudette) [inline Cargo.toml]: Reason for new dependencies: `rand`, `image`?

File: examples/text.rs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+impl nmg::Start for Demo {
+    fn start(
+        &mut self,
+        entities:   &mut entity::Manager,
+        components: &mut components::Container,
+    ) {
+        // Instantiate three entities
+        let object_0 = entities.add();
```
Comment (acgaudette) [inline examples/text.rs]: Text example probably doesn't actually need all the extra rendering

File: shaders/font3d.frag [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#version 450 core
+
+layout (location = 0) in vec2 inUV;
+
```
Comment (acgaudette) [inline shaders/font3d.frag]: Don't need this spacing here

File: src/lib.rs [hunk 7]
```diff
@@ -433,7 +436,10 @@ fn begin_update<T>(
         last_updated_renderer = now;
 
         // Render frame
-        if let Err(e) = context.draw(&components.draws.instances) {
+        if let Err(e) = context.draw(
```
Comment (acgaudette) [inline src/lib.rs]: You probably want
```
if let Err(e) = context.draw(
    &components.draws.instances,
    &mut components.text3ds,
) {
```

File: src/components/text3d.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
+extern crate fnv;
+
+use render;
+use entity;
+use components;
+
+use components::transform;
+use components::text;
+
+use font::*;
```
Comment (acgaudette) [inline src/components/text3d.rs]: I usually prefer not to do this, but maybe you have a good reason (`font::*`)

File: src/components/text3d.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    pub fn build(&mut self) -> TextBuilder {
+        TextBuilder::new(self)
+    }
+
+    fn set(&mut self, entity: entity::Handle, text: render::Text) {
+        debug_validate_entity!(self, entity);
+        *self.instances.get_mut(&entity).unwrap() = text;
+    }
+
+    // Update point light positions from transform component
```
Comment (acgaudette) [inline src/components/text3d.rs]: Comment?

File: src/components/text.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+) {
+    (**mapped) = render::FontData::new_raw(
+        x,
+        y,
+        z,
+        s,
+        t,
+        );
+    (*mapped) = (*mapped).offset(1);
+}
```
Comment (acgaudette) [inline src/components/text.rs]: The no EOLs in some of your files are a bit weird

File: src/components/text.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
+extern crate fnv;
```
Comment (acgaudette) [inline src/components/text.rs]: It is strange that this file is in `components`--should probably be one level up

File: src/components/text.rs [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+use render;
+use entity;
+
+use font::*;
+
+// TODO: Implement add_text functionality from render
```
Comment (acgaudette) [inline src/components/text.rs]: Is this comment still valid?

File: src/components/text.rs [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+// Add parameters here instead of just self
+pub fn prepare_bitmap_text(
```
Comment (acgaudette) [inline src/components/text.rs]: Some more verbose commenting would be helpful in this file

File: src/components/text3d.rs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    pub(crate) fn update(&mut self, transforms: &transform::Manager) {
+        for (entity, text3d) in &mut self.instances {
+            text3d.position = transforms.get_position(*entity);
+        }
+    }
+
+    //TODO: Implement add_text functionality from render
```
Comment (acgaudette) [inline src/components/text3d.rs]: Why is this fn wrapped again?

File: src/render.rs [hunk 15]
```diff
@@ -106,13 +110,16 @@ pub struct Context<'a> {
     _depth_image:     vd::Image,
     _views:           Vec<vd::ImageView>,
     _descriptor_pool: vd::DescriptorPool,
+    text_3d:          TextDisplay,
```
Comment (acgaudette) [inline src/render.rs]: The ordering is potentially strange here? (text after the persistent data)

File: src/render.rs [hunk 16]
```diff
@@ -106,13 +110,16 @@ pub struct Context<'a> {
...
+    font_data:        Font,
+    text_meta:        TextMeta,
 }
 
 impl<'a> Context<'a> {
     pub fn new(
         window:     &vdw::winit::Window,
         model_data: Vec<ModelData>,
-    ) -> vd::Result<Context> {
+    ) -> vd::Result<Context<'a>> {
```
Comment (acgaudette) [inline src/render.rs]: New lifetimes?

File: src/render.rs [hunk 17]
```diff
@@ -133,7 +140,7 @@ impl<'a> Context<'a> {
             _frag_mod,
             shader_stages,
         ) = load_shaders(device.clone())?;
-
+
```
Comment (acgaudette) [inline src/render.rs]: Watch out for this whitespace

File: src/render.rs [hunk 18]
```diff
@@ -218,6 +226,31 @@ impl<'a> Context<'a> {
         )?;
 
         let command_buffers = init_commands(&drawing_pool, &framebuffers)?;
+        let font_path = config::load_section_setting::<String>(
```
Comment (acgaudette) [inline src/render.rs]: This all probably belongs a bit higher up, with a `/* Text */` header or something (should be above debug and the command buffers at least)

File: src/render.rs [hunk 19]
```diff
@@ -341,9 +378,7 @@ impl<'a> Context<'a> {
 
         /* Coup */
 
-        self.swapchain = swapchain;
         self.command_fences = command_fences;
-        self.render_pass = render_pass;
```
Comment (acgaudette) [inline src/render.rs]: It looks like you've made some changes to the base `render.rs` along with the new text code. If you could split up any changes to a separate commit, that would be great.

File: src/render.rs [hunk 20]
```diff
@@ -1218,9 +1306,18 @@ fn init_vulkan(window: &vdw::winit::Window) -> vd::Result<(
 
     /* Instance */
 
+    //TODO: Change to setting for use with RenderDoc
```
Comment (acgaudette) [inline src/render.rs]: ? Artifact

File: src/render.rs [hunk 21]
```diff
@@ -2426,7 +2525,6 @@ fn init_drawing(
         .offset(0)
         .range(shared_alignment)
         .build();
-
```
Comment (acgaudette) [inline src/render.rs]: Try to remove diffs like this which are just changing the original file

File: src/render.rs [hunk 22]
```diff
@@ -2697,3 +2798,793 @@ unsafe fn copy_buffer<T: std::marker::Copy>(
...
+    let viewport = vd::Viewport::builder()
+        .x(0.0f32)
+        .y(0.0f32)
+        .width(extent2d.width() as f32)
+        .height(extent2d.height() as f32)
+        .min_depth(0.0f32)
+        .max_depth(1.0f32)
+        .build();
+        
+    //eprintln!("Width: {}", extent2d.width());
```
Comment (acgaudette) [inline src/render.rs]: ?

File: src/render.rs [hunk 23]
```diff
@@ -2697,3 +2798,793 @@ unsafe fn copy_buffer<T: std::marker::Copy>(
...
+            "settings",
+            "shader_path"
+        );
+
+        [path, "/"].concat()
+    };
+
+    let type_shader = if is_3d {"3"} else {"2"};
+    
+    println!("Loading {}d text shaders from \"{}\"", type_shader, path);
```
Comment (acgaudette) [inline src/render.rs]: If you're going to add this message, we should probably add one for the debug shaders as well

File: src/render.rs [hunk 24]
```diff
@@ -2697,3 +2798,793 @@ unsafe fn copy_buffer<T: std::marker::Copy>(
...
+            host_buffer,
+            _image.handle(),
+            vd::ImageLayout::TransferDstOptimal,
+            &regions
+        );
+    }
+
+    copy_cmd.end()?;
+
+    //TODO: Clean up buffers
```
Comment (acgaudette) [inline src/render.rs]: Watch out

**Review by acgaudette — CHANGES_REQUESTED** [review]
See comments.
Need `text.rs` to be documented more until I can really comment.
Did a quick pass of render, in general looks good but seems like you still have a number of TODOs floating around. Make sure you're taking care of memory and concurrency issues.

File: src/lib.rs (hunk 7 above)
Comment (hobogenized) [reply src/lib.rs]: Done

File: src/components/text.rs [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+
+use font::*;
+
+/**
+ * Shared method for preparing and rendering text isntances, whether 3d or not
+ */
+pub fn prepare_bitmap_text(
+    instances: &mut fnv::FnvHashMap<entity::Handle, render::Text>,
+    font_data: &Font,
+    ptr: *mut *mut render::FontData,
```
Comment (acgaudette) [inline src/components/text.rs]: Do both of these pointers need to be mutable?

**Comment by acgaudette** [comment]
Getting error: `DEBUG_REPORT:  [ UNASSIGNED-CoreValidation-Shader-DescriptorTypeMismatch ] Object: VK_NULL_HANDLE (Type = 0) | Type mismatch on descriptor slot 0.1 (expected 'VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER') but descriptor of type VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC` @hobogenized

**Review by acgaudette — APPROVED** [review]

**Pull request closed without merging**